src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -1621,6 +1650,13 @@ private bool ExactPointerInference(TypeWithAnnotations source, TypeWithAnnotatio
                      target.Type is FunctionPointerTypeSymbol { Signature: { ParameterCount: int targetParameterCount } targetSignature } &&
                      sourceParameterCount == targetParameterCount)
             {
+                if (sourceSignature.RefKind != targetSignature.RefKind
+                    || sourceSignature.ParameterRefKinds != targetSignature.ParameterRefKinds
+                    || !callingConventionsEqual(sourceSignature, targetSignature))
```

Comment:
>!callingConventionsEqual(sourceSignature, targetSignature) [](start = 23, length = 58)

Should we do the same check for upper/lower bound inference?

Reply:
Upper and lower bound inference do actual overload resolution, which handles refkind and calling convention mismatches.

---
In reply to: [562997572](https://github.com/dotnet/roslyn/pull/50249#discussion_r562997572) [](ancestors = 562997572)

Reply:
> Upper and lower bound inference do actual overload resolution, which handles refkind and calling convention mismatches.

I do not see LowerBoundFunctionPointerTypeInference and UpperBoundFunctionPointerTypeInference do overload resolution. What am I missing.

---
In reply to: [563941011](https://github.com/dotnet/roslyn/pull/50249#discussion_r563941011) [](ancestors = 563941011,562997572)

Reply:
> I do not see LowerBoundFunctionPointerTypeInference and UpperBoundFunctionPointerTypeInference do overload resolution. What am I missing?

It's done in phase 2 of inference, whereas this is all part of phase 1. I could add the checks here as well, but there won't be an observable output change.

---
In reply to: [564004088](https://github.com/dotnet/roslyn/pull/50249#discussion_r564004088) [](ancestors = 564004088,563941011,562997572)

Reply:
> It's done in phase 2 of inference, whereas this is all part of phase 1. 

This doesn't quite explain anything for me.

> I could add the checks here as well, but there won't be an observable output change.

This sounds suspicious to me. Could you please elaborate, pointing where the check would be performed instead?

---
In reply to: [564009761](https://github.com/dotnet/roslyn/pull/50249#discussion_r564009761) [](ancestors = 564009761,564004088,563941011,562997572)

Reply:
Hmm, actually, I believe in my testing I wasn't hitting the codepath because I was using address of method groups, which need to perform overload resolution (and will fail to do so in the presence of mismatches). I can hit it with variables of the appropriate pointer type, I'll add more testing.

---
In reply to: [564016180](https://github.com/dotnet/roslyn/pull/50249#discussion_r564016180) [](ancestors = 564016180,564009761,564004088,563941011,562997572)

Reply:
Tests added.

---
In reply to: [564040602](https://github.com/dotnet/roslyn/pull/50249#discussion_r564040602) [](ancestors = 564040602,564016180,564009761,564004088,563941011,562997572)